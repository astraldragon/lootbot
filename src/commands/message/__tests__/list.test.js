jest.mock("discord-akairo");

jest.mock("../../../models", () => ({
  Message: { findAll: jest.fn("findAll") },
  Tier: { findAll: jest.fn("findAll") }
}));

const { Message, Tier } = require("../../../models");
const { checkManagePermissions } = require("../../../support");
const { Command } = require("discord-akairo");
const ListCommand = require("../list");

describe("commands/loot/list", () => {
  beforeEach(() => {
    this.command = new ListCommand();

    this.msg = {
      guild: { id: "1234" },
      channel: {
        send: jest.fn("send").mockReturnValue(Promise.resolve())
      }
    };
  });

  it("should configure properly", () => {
    expect(Command).toMatchSnapshot();
  });

  it("should have the proper permission check", () => {
    expect(Command.mock.calls[0][1].options.permissions).toBe(
      checkManagePermissions
    );
  });

  it("should fetch the messages for the guild", () => {
    this.command.exec(this.msg);

    expect(Message.findAll).toHaveBeenCalledWith({
      include: [{ model: Tier }],
      where: { guild: this.msg.guild.id }
    });
  });

  it("should notify when no loot is found for the guild", async () => {
    Message.findAll.mockReturnValue([]);
    await this.command.exec(this.msg);
    expect(this.msg.channel.send).toHaveBeenCalledWith("No messages found");
  });

  it("should list the messages", async () => {
    Message.findAll.mockReturnValue([
      {
        name: "msg1",
        message: "this is a message",
        type: "intro",
        Tier: null,
        user_id: null
      },
      {
        name: "msg1",
        message: "this is a message",
        type: "intro",
        Tier: { name: "Common" },
        user_id: "1234"
      }
    ]);

    await this.command.exec(this.msg);
    expect(this.msg.channel.send).toMatchSnapshot();
  });

  it("should notify if any errors occur", async () => {
    Message.findAll.mockImplementation(() => {
      throw new Error("test");
    });

    await this.command.exec(this.msg);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      "An error occurred listing messages"
    );
  });
});
