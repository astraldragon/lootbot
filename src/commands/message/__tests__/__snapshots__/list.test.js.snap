// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/loot/list should configure properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "message:list",
      Object {
        "aliases": Array [
          "message-list",
          "ml",
        ],
        "category": "Message",
        "channelRestriction": "guild",
        "description": Object {
          "content": "List the glorious messages",
          "examples": Array [
            "message-list",
          ],
        },
        "options": Object {
          "permissions": [Function],
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`commands/loot/list should list the messages 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "
__**msg1**__
Message: this is a message
Type: intro
Tier: N/A
User: N/A

__**msg1**__
Message: this is a message
Type: intro
Tier: Common
User: 1234
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;
